src/transform/src/analysis/equivalences.rs

```diff
@@ -714,4 +718,122 @@ impl EquivalenceClasses {
+                simplified = self.reduce_expr(expr1) || simplified;
+                simplified = self.reduce_expr(expr2) || simplified;
+            }
+            MirScalarExpr::CallUnary { expr, .. } => {
+                simplified = self.reduce_expr(expr) || simplified;
+            }
+            MirScalarExpr::CallVariadic { exprs, .. } => {
+                for expr in exprs.iter_mut() {
+                    simplified = self.reduce_expr(expr) || simplified;
+                }
+            }
+            MirScalarExpr::If { cond: _, then, els } => {
+                // Do not simplify `cond`, as we cannot ensure the simplification
+                // continues to hold as expressions migrate around.
+                simplified = self.reduce_expr(then) || simplified;
+                simplified = self.reduce_expr(els) || simplified;
+            }
+            _ => {}
+        }
+        simplified
+    }
+}
+
+impl ExpressionReducer for BTreeMap<&MirScalarExpr, &MirScalarExpr> {
+    /// Perform any exact replacement for `expr`, report if it had an effect.
+    fn replace(&self, expr: &mut MirScalarExpr) -> bool {
+        if let Some(other) = self.get(expr) {
+            if other != &expr {
+                expr.clone_from(other);
+                return true;
+            }
+        }
+        false
+    }
+}
+
+impl ExpressionReducer for BTreeMap<MirScalarExpr, MirScalarExpr> {
+    /// Perform any exact replacement for `expr`, report if it had an effect.
+    fn replace(&self, expr: &mut MirScalarExpr) -> bool {
+        if let Some(other) = self.get(expr) {
+            if other != expr {
+                expr.clone_from(other);
+                return true;
+            }
+        }
+        false
+    }
+}
+
+trait UnionFind<T> {
+    /// Sets `self[x]` to the root from `x`, and returns a reference to the root.
+    fn find<'a>(&'a mut self, x: &T) -> Option<&'a T>;
+    /// Ensures that `x` and `y` have the same root.
+    fn union(&mut self, x: &T, y: &T);
+}
+
+impl<T: Clone + Ord> UnionFind<T> for BTreeMap<T, T> {
+    fn find<'a>(&'a mut self, x: &T) -> Option<&'a T> {
+        if !self.contains_key(x) {
+            None
+        } else {
+            if self[x] != self[&self[x]] {
+                // Path halving
+                let mut y = self[x].clone();
+                while y != self[&y] {
+                    let grandparent = self[&self[&y]].clone();
+                    *self.get_mut(&y).unwrap() = grandparent;
+                    y.clone_from(&self[&y]);
+                }
+                *self.get_mut(x).unwrap() = y;
+            }
+            Some(&self[x])
+        }
+    }
+
+    fn union(&mut self, x: &T, y: &T) {
+        match (self.find(x).is_some(), self.find(y).is_some()) {
+            (true, true) => {
+                if self[x] != self[y] {
+                    let root_x = self[x].clone();
+                    let root_y = self[y].clone();
+                    self.insert(root_x, root_y);
```

Comment:
Yup, in that discussion it says that just doing path compression already gives you n log n time. Adding in either of the weighted union approaches brings it down to inverse Ackermann, but they aren't necessary if n log n is good enough.

Reply:
But could you please add a comment in the code noting that this is an NlogN version of Union-Find? (So that the next reader of the code won't start wondering about the same thing.)